Source/WebCore/dom/ContainerNodeAlgorithms.cpp

```diff
@@ -101,38 +101,47 @@ inline void updateObservability(RemovedSubtreeObservability& currentObservabilit
+static RemovedSubtreeResult notifyNodeRemovedFromDocument(ContainerNode& oldParentOfRemovedTree, TreeScopeChange treeScopeChange, Node& node)
 {
     ASSERT(!node.parentNode());
     ASSERT(oldParentOfRemovedTree.isConnected());
     ASSERT(node.isConnected());
 
     RemovedSubtreeObservability observability = RemovedSubtreeObservability::NotObservable;
+    auto childrenDeletionCanBeDelayed = ContainerNode::ChildrenDeletionCanBeDelayed::Yes;
+    int subTreeSize = 0;
     for (RefPtr currentNode = &node; currentNode; currentNode = NodeTraversal::next(*currentNode)) {
+        subTreeSize++;
         currentNode->removedFromAncestor(Node::RemovalType { /* disconnectedFromDocument */ true, treeScopeChange == TreeScopeChange::Changed }, oldParentOfRemovedTree);
+        if (!AsyncNodeDeletionQueue::canNodeBeAsyncDeleted(node))
+            childrenDeletionCanBeDelayed = ContainerNode::ChildrenDeletionCanBeDelayed::No;
         updateObservability(observability, observabilityOfRemovedNode(*currentNode));
         if (RefPtr root = currentNode->shadowRoot())
-            updateObservability(observability, notifyNodeRemovedFromDocument(oldParentOfRemovedTree, TreeScopeChange::DidNotChange, *root));
+            updateObservability(observability, notifyNodeRemovedFromDocument(oldParentOfRemovedTree, TreeScopeChange::DidNotChange, *root).removedSubtreeObservability);
     }
-    return observability;
+    return { subTreeSize, observability, childrenDeletionCanBeDelayed };
 }
 
-static RemovedSubtreeObservability notifyNodeRemovedFromTree(ContainerNode& oldParentOfRemovedTree, TreeScopeChange treeScopeChange, Node& node)
+static RemovedSubtreeResult notifyNodeRemovedFromTree(ContainerNode& oldParentOfRemovedTree, TreeScopeChange treeScopeChange, Node& node)
 {
     ASSERT(!node.parentNode());
     ASSERT(!oldParentOfRemovedTree.isConnected());
-
+    int subTreeSize = 0;
+    auto childrenDeletionCanBeDelayed = ContainerNode::ChildrenDeletionCanBeDelayed::Yes;
     RemovedSubtreeObservability observability = RemovedSubtreeObservability::NotObservable;
     for (RefPtr currentNode = &node; currentNode; currentNode = NodeTraversal::next(*currentNode)) {
+        subTreeSize++;
         currentNode->removedFromAncestor(Node::RemovalType { /* disconnectedFromDocument */ false, treeScopeChange == TreeScopeChange::Changed }, oldParentOfRemovedTree);
+        if (!AsyncNodeDeletionQueue::canNodeBeAsyncDeleted(node))
+            childrenDeletionCanBeDelayed = ContainerNode::ChildrenDeletionCanBeDelayed::No;
         updateObservability(observability, observabilityOfRemovedNode(*currentNode));
         if (RefPtr root = currentNode->shadowRoot())
-            updateObservability(observability, notifyNodeRemovedFromTree(oldParentOfRemovedTree, TreeScopeChange::DidNotChange, *root));
+            updateObservability(observability, notifyNodeRemovedFromTree(oldParentOfRemovedTree, TreeScopeChange::DidNotChange, *root).removedSubtreeObservability);
```

Comment:
I don't think we should ignore the number of nodes or whether nodes can be deleted async or not in shadow trees.